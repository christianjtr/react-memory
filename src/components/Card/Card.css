.memory-card {
  -webkit-box-shadow: 0px 0px 1px 0px rgba(128, 128, 128, 1);
  -moz-box-shadow: 0px 0px 1px 0px rgba(128, 128, 128, 1);
  box-shadow: 0px 0px 1px 0px rgba(128, 128, 128, 1);
  border-radius: 50px;
  height: 100%;
}

.memory-card > .memory-card__image {
  border-radius: inherit;
}

.memory-card > .memory-card__overlay {
  background: LightSteelBlue;
  height: inherit;
}

.memory-card__overlay:hover {
  border: 5px solid CadetBlue;
}

.memory-card__overlay img {
  opacity: 0;
}
