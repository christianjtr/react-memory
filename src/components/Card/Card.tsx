import React from 'react';
import './Card.css';
import { GameCard } from '../../types';

interface CardProps {
  data: GameCard;
  onClick: (gameCardId: number, uniqueId: number) => void;
}

const Card: React.FC<CardProps> = (props: CardProps): React.ReactElement => {
  const {
    data: {
      data: { id, loginName, avatarURL },
      id: gameCardId,
      isFaceDown,
    },
    onClick: onClickCardCallback,
  } = props;

  return (
    <div
      className="memory-card border-solid border-1 border-grey-200 rounded-lg"
      tabIndex={0}
      role="button"
      aria-pressed={false}
      title="Click to flip the card"
      onClick={(): void => onClickCardCallback(gameCardId, id)}>
      <div
        className={`memory-card__image rounded-lg ${
          isFaceDown && 'memory-card__overlay'
        }`}>
        <img src={avatarURL} alt={loginName} className="rounded-lg" />
      </div>
    </div>
  );
};

export default Card;
