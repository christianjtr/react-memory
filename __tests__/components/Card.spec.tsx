import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, cleanup, act, fireEvent } from '@testing-library/react';
import { createRenderer } from 'react-test-renderer/shallow';

import Card, { CardProps } from '../../src/components/Card/Card';

describe('Card component', () => {
  const onClickCardCallback = jest.fn();

  const faceDownCardData: CardProps = {
    data: {
      data: {
        id: 1,
        loginName: 'Contributor',
        avatarURL: 'dummy-url',
      },
      isFaceDown: true,
      id: 12345,
    },
    onClick: onClickCardCallback,
    className: 'dummy-class',
  };
  const faceUpCardData: CardProps = {
    data: {
      data: {
        id: 1,
        loginName: 'Contributor',
        avatarURL: 'dummy-url',
      },
      isFaceDown: false,
      id: 12345,
    },
    onClick: onClickCardCallback,
    className: 'dummy-class',
  };

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it('should render the component correctly', () => {
    const renderer = createRenderer();
    const component = renderer.render(<Card {...faceDownCardData} />);

    expect(component).toBeDefined();
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it('should render the image element correctly - faceDown', () => {
    const { container } = render(<Card {...faceDownCardData} />);

    const imageContainer = container.getElementsByClassName('memory-card__overlay');
    const image = screen.getByRole('img');

    expect(imageContainer).toBeDefined();
    expect(imageContainer.length).toBe(1);

    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', faceDownCardData.data.data.avatarURL);
    expect(image).toHaveAttribute('alt', faceDownCardData.data.data.loginName);
  });

  it('should render the image element correctly - faceUp', () => {
    const { container } = render(<Card {...faceUpCardData} />);

    const imageContainer = container.getElementsByClassName('memory-card__overlay');
    const image = screen.getByRole('img');

    expect(imageContainer.length).toBe(0);
    expect(image).toBeInTheDocument();
  });

  it('should call the event handler when clicking on the memory card', () => {
    render(<Card {...faceDownCardData} />);

    const divButtonElement = screen.getByRole('button');

    expect(divButtonElement).toBeDefined();
    expect(typeof divButtonElement.onclick === 'function').toBeTruthy();

    act(() => {
      fireEvent.click(divButtonElement);
    });

    expect(onClickCardCallback).toHaveBeenCalled();
  });
});
