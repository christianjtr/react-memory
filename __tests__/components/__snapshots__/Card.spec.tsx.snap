// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component should render the component correctly 1`] = `
<div
  aria-pressed={false}
  className="memory-card border-solid border-1 border-grey-200 rounded-lg dummy-class"
  onClick={[Function]}
  role="button"
  tabIndex={0}
  title="Click to flip the card"
>
  <div
    className="memory-card__image rounded-lg memory-card__overlay"
  >
    <img
      alt="Contributor"
      className="rounded-lg"
      src="dummy-url"
    />
  </div>
</div>
`;
